### ISSUE TITLE
Proposal to add ExtraProps to the configuration

### ISSUE THREAD

**Issue by Burgyn** [ThreeMammals/Ocelot#1990]
If the need arises on the Ocelot usage side for an extension, we usually have to work with some custom settings in the Ocelot downstream route configuration as well.

The current options are that we put the custom settings into a custom setup and use the downstream key as a reference to it, or we can add them to the original JSON setup and read them ourselves (duplicate read).

From an extensibility point of view, it would be nice if we could add additional properties that the programmer could then read.

## Proposal

```json
 "ExtraProps" : {
    "CachePolicies": ["policy1", "policy2"],
    "SwaggerKey": "todo"
  }
```

### Whole usage

```json
{
    "Routes": [
        {
          "DownstreamPathTemplate": "/todos/{id}",
          "DownstreamScheme": "https",
          "DownstreamHostAndPorts": [
            {
               "Host": "jsonplaceholder.typicode.com",
               "Port": 443
            }
          ],
         "UpstreamPathTemplate": "/todos/{id}",
         "UpstreamHttpMethod": [ "Get" ],
         // 👇 additional properties for better extensibility
         "ExtraProps" : {
            "CachePolicies": ["policy1", "policy2"],
            "SwaggerKey": "todo",
            "AnotherProp": "value"
          }
        }
    ]
}
```
Usage in middleware:

```csharp
PreQueryStringBuilderMiddleware = async (context, next) =>
{
  DownstreamRoute? route = context.Items.DownstreamRoute();
  var cachePolicies = route.GetExtraProps<string[]>("CachePolicies");
  ...
}
```

I can prepare PR for this change. I just want to check if this is a change that fits the direction of the project.

Thank you in advance for your reply.

---

```json
{
    "Routes": [
        {
          "DownstreamPathTemplate": "/todos/{id}",
          "DownstreamScheme": "https",
          "DownstreamHostAndPorts": [
            {
               "Host": "jsonplaceholder.typicode.com",
               "Port": 443
            }
          ],
         "UpstreamPathTemplate": "/todos/{id}",
         "UpstreamHttpMethod": [ "Get" ],
         // 👇 additional properties for better extensibility
         "CachePolicies": ["policy1", "policy2"],
         "SwaggerKey": "todo",
         "AnotherProp": "value"
        }
    ]
}
```

> Theoretically, we can also do that extra properties will be on the main level. This is nicer from a usage point of view, but more complicated to implement.
> I don't know what you prefer.

**Comment by raman-m** [comment]
Duplicate of #738

**Comment by Burgyn** [comment]
Hi @raman-m,

thanks for linking.

M.

**Comment by raman-m** [comment]
Reopen for official closing

**Comment by raman-m** [comment]
Closed by #1843
